plugins/optimization-detective/detect.js

```diff
@@ -129,12 +112,131 @@ function getCurrentTime() {
 	return Date.now();
 }
 
+/**
+ * Recursively freezes an object to prevent mutation.
+ *
+ * @param {Object} obj Object to recursively freeze.
+ */
+function recursiveFreeze( obj ) {
+	for ( const prop of Object.getOwnPropertyNames( obj ) ) {
+		const value = obj[ prop ];
+		if ( null !== value && typeof value === 'object' ) {
+			recursiveFreeze( value );
+		}
+	}
+	Object.freeze( obj );
+}
+
+/**
+ * URL metric being assembled for submission.
+ *
+ * @type {URLMetric}
+ */
+let urlMetric;
+
+/**
+ * Reserved root property keys.
+ *
+ * @see {URLMetric}
+ * @see {AmendedElementData}
+ * @type {Set<string>}
+ */
+const reservedRootPropertyKeys = new Set( [ 'url', 'viewport', 'elements' ] );
+
+/**
+ * Gets root URL Metric data.
+ *
+ * @return {URLMetric} URL Metric.
+ */
+function getRootData() {
+	const immutableUrlMetric = structuredClone( urlMetric );
+	recursiveFreeze( immutableUrlMetric );
+	return immutableUrlMetric;
+}
+
+/**
+ * Amends root URL metric data.
+ *
+ * @todo Would "extend" be better than "amend"? Or something else?
```

Comment:
Or `setRootData`?

Reply:
Or not, as that implies it is overriding all of the root data which it doesn't allow.

Reply:
amend seems fine, although extend is more familiar WordPress parlance

Reply:
Renamed in 73d2252961228c07f8ec1e2af82fd1cd588bb09d